services/builder/main.go

```diff
@@ -32,23 +32,53 @@ const (
 
 // healthCheck function is function used for querying readiness of the pod running this microservice
 func healthCheck(usecases *usecases.Usecases) func() error {
-	return func() error {
-		if usecases.Terraformer.PerformHealthCheck() != nil {
-			return errors.New("terraformer is unhealthy")
-		}
-		if usecases.Ansibler.PerformHealthCheck() != nil {
-			return errors.New("ansibler is unhealthy")
-		}
-		if usecases.KubeEleven.PerformHealthCheck() != nil {
-			return errors.New("kube-eleven is unhealthy")
+	type HealthCheck struct {
+		timeSinceTerraformFailure  *time.Time
+		timeSinceAnsiblerFailure   *time.Time
+		timeSinceKubeElevenFailure *time.Time
+		timeSinceKuberFailure      *time.Time
+		timeSinceContextBoxFailure *time.Time
+	}
+	updateTimeSinceFailure := func(now *time.Time, t **time.Time, err error) {
+		if err == nil {
+			*t = nil
+			return
 		}
-		if usecases.Kuber.PerformHealthCheck() != nil {
-			return errors.New("kuber is unhealthy")
+		if *t == nil {
+			*t = now
 		}
-		if usecases.ContextBox.PerformHealthCheck() != nil {
-			return errors.New("context-box is unhealthy")
+	}
+	healthCheck := func(check *HealthCheck) {
+		now := time.Now()
+		updateTimeSinceFailure(&now, &check.timeSinceTerraformFailure, usecases.Terraformer.PerformHealthCheck())
+		updateTimeSinceFailure(&now, &check.timeSinceAnsiblerFailure, usecases.Ansibler.PerformHealthCheck())
+		updateTimeSinceFailure(&now, &check.timeSinceKubeElevenFailure, usecases.KubeEleven.PerformHealthCheck())
+		updateTimeSinceFailure(&now, &check.timeSinceKuberFailure, usecases.Kuber.PerformHealthCheck())
+		updateTimeSinceFailure(&now, &check.timeSinceContextBoxFailure, usecases.ContextBox.PerformHealthCheck())
+	}
+	checkFailure := func(t *time.Time, service string, perr error) error {
+		if t != nil && time.Since(*t) >= 4*time.Minute {
```

Comment:
Isn't 4 minutes bit to long? If the healtheck is performed every 30s, it will require 8 failures.

Reply:
I think 4 mins is not high. The connectivity problem is not directly linked to the health of the service. In most cases, the connectivity status should not be affected by a service restart at all.

Reply:
I think 4mins is enough, I've tested this and it takes just >90secs to restart terraformer